gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+                if secs < 30 {
+                    slog::debug!(
+                        &self.log,
+                        "waiting for SP discovery to complete...";
+                        "elapsed" => ?elapsed,
+                        "error" => err,
+                    );
+                } else if secs < 180 {
+                    slog::info!(
+                        &self.log,
+                        "still waiting for SP discovery to complete...";
+                        "elapsed" => ?elapsed,
+
+                        "error" => err,
+                    )
+                } else {
+                    slog::warn!(
+                        &self.log,
+                        "we have been waiting for SP discovery to complete \
+                         for a pretty long time!";
+                        "elapsed" => ?elapsed,
+                        "error" => err,
+                    )
+                }
+            },
+        )
+        .await
+        .expect("we should never return a fatal error here");
+
+        slog::info!(
+            &self.log,
+            "starting to polling SP sensor data every {:?}", self.poll_interval;
+        );
+
+        let mut known_sps = HashMap::with_capacity(NORMAL_NUMBER_OF_SPS);
+        for (spid, _) in sps {
+            slog::info!(
+                &self.log,
+                "found a new little friend!";
+                "sp_slot" => ?spid.slot,
+                "chassis_type" => ?spid.typ,
+            );
+            let task = self.tasks.spawn(self.make_poller(spid, rack_id));
+            let _prev = known_sps.insert(spid, task);
+            debug_assert!(
+                _prev.is_none(),
+                "we should not have clobbered an existing SP poller!"
+            );
+        }
+
+        // All pollers started! Now wait to see if any of them have died...
+        loop {
+            let mut joined = self.tasks.join_next().await;
+            while let Some(result) = joined {
+                if let Err(e) = result {
+                    if cfg!(debug_assertions) {
+                        unreachable!(
+                            "we compile with `panic=\"abort\"`, so a spawned task \
+                             panicking should abort the whole process..."
+                        );
+                    } else {
+                        slog::error!(
+                            &self.log,
+                            "a spawned SP poller task panicked! this should \
+                             never happen: we compile with `panic=\"abort\"`, so \
+                             a spawned task panicking should abort the whole \
+                             process...";
+                            "error" => %e,
+                        );
+                    }
+                }
+                // drain any remaining errors
+                joined = self.tasks.try_join_next();
+            }
+
+            for (spid, poller) in &mut known_sps {
```

Comment:
I feel like I'm missing something about this section. It looks like the poller tasks can return an error if they fail to reach an SP. They don't panic, which makes me think I don't understand the `unreachable!()` above either. I think I'm confused about the situations in which one would have a poller task return (not panic) and where we want to restart it. Hopefully my question makes sense!

Reply:
Yes, I'm in the process of reworking this code --- it's due to a misunderstanding of which errors the `ManagementSwitch` functions will return and in what situations. It turns out that looking up a SP by ID only fails if discovery hasn't run yet, or if the ID is totally invalid, which are (a) fatal errors and (b) should never happen due to the poller tasks only being started once discovery completes. So, this code is  about to get a lot simpler; there should never be a reason to restart the poller tasks. See this conversation with @jgallagher: https://github.com/oxidecomputer/omicron/pull/6354#discussion_r1725727261

Reply:
...and 8a6252ab979d79d5297bda9b2d1b786874359e5b has completely obviated all of this. :)